src/dataflow/src/source/postgres.rs

```diff
@@ -0,0 +1,319 @@
+        self.packer.clear();
+
+        for (val, ty) in tuple.tuple_data().iter().zip(col_types.iter()) {
+            let datum = match val {
+                TupleData::Null => Datum::Null,
+                TupleData::Toast => Datum::Null, // FIXME
+                TupleData::Text(b) => {
+                    let s = std::str::from_utf8(&b).unwrap();
+                    match ty.scalar_type {
+                        PgType::BOOL => parse_datum(s, ScalarType::Bool).unwrap(),
+                        PgType::INT4 => parse_datum(s, ScalarType::Int32).unwrap(),
+                        PgType::INT8 => parse_datum(s, ScalarType::Int64).unwrap(),
+                        PgType::FLOAT4 => parse_datum(s, ScalarType::Float32).unwrap(),
+                        PgType::FLOAT8 => parse_datum(s, ScalarType::Float64).unwrap(),
+                        PgType::DATE => parse_datum(s, ScalarType::Date).unwrap(),
+                        PgType::TIME => parse_datum(s, ScalarType::Time).unwrap(),
+                        PgType::TIMESTAMP => parse_datum(s, ScalarType::Timestamp).unwrap(),
+                        PgType::TEXT => parse_datum(s, ScalarType::String).unwrap(),
+                        PgType::UUID => parse_datum(s, ScalarType::Uuid).unwrap(),
+                        ref other => todo!("Unsupported data type {:?}", other),
+                    }
+                }
+            };
+            self.packer.push(datum);
+        }
+
+        self.packer.finish_and_reuse()
+    }
+
+    async fn into_generator(mut self, timestamper: Timestamper, mut co: Co<StreamItem>) {
+        let (rel_id, col_types) = self.table_info().await;
+
+        let conninfo = &self.connector.conn;
+        let (mut client, conn) = connect_replication(&conninfo, NoTls, ReplicationMode::Logical)
+            .await
+            .unwrap();
+
+        // TODO communicate errors back into the stream
+        tokio::spawn(async move {
+            if let Err(e) = conn.await {
+                eprintln!("connection error: {}", e);
+            }
+        });
+
+        let (slot_name, mut cur_lsn) = self
+            .produce_snapshot(&mut client, &col_types, &timestamper, &mut co)
+            .await;
+
+        // We now switch to consuming the stream
+        let mut stream = client
+            .start_logical_replication(&slot_name, cur_lsn, &self.plugin)
+            .await
+            .unwrap();
+
+        let mut time_lease = None;
+        while let Some(item) = stream.next().await {
+            match item {
+                Ok(ReplicationMessage::XLogData(xlog_data)) => {
+                    use LogicalReplicationMessage::*;
+
+                    match xlog_data.data() {
+                        Begin(_) => {
+                            // Lease a tx time
+                            time_lease = Some(timestamper.lease().await);
+                        }
+                        Insert(insert) if insert.rel_id() == rel_id => {
```

Comment:
i just want to double check my understanding here -- if we want to replicate 5 tables from upstream we will be reading in 5 copies of the pg log, and filtering for the appropriate table right?

Reply:
With this PR, yes. The next step is to implement the `CREATE SOURCES` (note the plural) syntax that will create all sources at once. I'm hoping to do some magic during that codepath that gives some way for those sources to coordinate and consume a single stream.

I plan to use the same coordination mechanism to make sure I assign timestamps consistently between the various tables (related to your main comment above)